File: ios/Hapi/Features/Chat/Blocks/ToolCallBlockView.swift
import HapiClient
import HapiProtocol
import HapiUI
import SwiftUI

/// One tool invocation (web `ToolCard`): collapsed header row — icon, title,
/// subtitle, status — expanding to the per-tool body (`ToolCallBody`), the
/// permission state, and nested children (sidechain transcript behind an
/// indent rail). Cards with a pending permission start expanded and carry the
/// "awaiting approval" banner; with a `\.chatInteractions` engine present
/// (A-M3b) the banner grows the actionable approval footer.
struct ToolCallBlockView: View {
    let block: ToolCallBlock
    let basePath: String?

    @State private var expanded: Bool
    @State private var childrenOpen: Bool
    @Environment(\.hapiTheme) private var theme
    @Environment(\.chatInteractions) private var interactions

    init(block: ToolCallBlock, basePath: String?) {
        self.block = block
        self.basePath = basePath
        let pending = block.tool.permission?.status == .pending
        _expanded = State(initialValue: pending)
        _childrenOpen = State(initialValue: pending)
    }

    var body: some View {
        let presentation = toolCardPresentation(block.tool, basePath: basePath)
        VStack(alignment: .leading, spacing: 0) {
            headerRow(presentation)
            if let permission = block.tool.permission {
                if permission.status == .pending, let interactions {
                    pendingApprovalSection(permission: permission, interactions: interactions)
                } else {
                    PermissionStateRow(permission: permission)
                }
            }
            if expanded {
                ToolCallBody(tool: block.tool, basePath: basePath)
                    .padding(.horizontal, 10)
                    .padding(.bottom, 10)
            }
            if !block.children.isEmpty {
                childrenSection
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(theme.surface)
        .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
    }

    /// Pending permission with a live interaction engine: highlighted banner
    /// plus the actionable footer (approve buttons / answer forms).
    private func pendingApprovalSection(
        permission: ToolPermission,
        interactions: ChatInteractor
    ) -> some View {
        VStack(alignment: .leading, spacing: 0) {
            Label("Awaiting approval", systemImage: "hourglass")
                .font(.footnote.weight(.semibold))
                .foregroundStyle(.orange)
                .padding(.horizontal, 10)
                .padding(.top, 6)
            PendingPermissionFooter(
                tool: block.tool,
                requestId: permission.id,
                interactions: interactions
            )
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(.orange.opacity(0.10))
    }

    private func headerRow(_ presentation: ToolCardPresentation) -> some View {
        Button {
            withAnimation(.easeInOut(duration: 0.15)) {
                expanded.toggle()
            }
        } label: {
            HStack(spacing: 8) {
                Image(systemName: presentation.icon)
                    .font(.footnote)
                    .foregroundStyle(.secondary)
                    .frame(width: 18)
                VStack(alignment: .leading, spacing: 1) {
                    Text(presentation.title)
                        .font(.subheadline)
                        .foregroundStyle(theme.textPrimary)
                        .lineLimit(1)
                    if let subtitle = presentation.subtitle {
                        Text(subtitle)
                            .font(.caption2)
                            .foregroundStyle(.secondary)
                            .lineLimit(1)
                    }
                }
                Spacer(minLength: 8)
                ToolStatusIndicator(state: block.tool.state)
            }
            .padding(.horizontal, 10)
            .padding(.vertical, 8)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    /// Sidechain children, nested behind an indent rail; collapsed to a
    /// count row.
    private var childrenSection: some View {
        VStack(alignment: .leading, spacing: 0) {
            Button {
                withAnimation(.easeInOut(duration: 0.15)) {
                    childrenOpen.toggle()
                }
            } label: {
                Label(
                    block.children.count == 1
                        ? String(localized: "1 agent step")
                        : String(format: String(localized: "%lld agent steps"), Int64(block.children.count)),
                    systemImage: childrenOpen ? "chevron.down" : "chevron.right"
                )
                .font(.caption)
                .foregroundStyle(.secondary)
                .padding(.horizontal, 10)
                .padding(.vertical, 6)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            if childrenOpen {
                HStack(alignment: .top, spacing: 10) {
                    RoundedRectangle(cornerRadius: 1)
                        .fill(theme.divider)
                        .frame(width: 2)
                    VStack(alignment: .leading, spacing: 8) {
                        ForEach(block.children, id: \.id) { child in
                            ChatSubBlockView(block: child, basePath: basePath)
                        }
                    }
                }
                .padding(.leading, 12)
                .padding(.trailing, 8)
                .padding(.bottom, 10)
            }
        }
    }
}

// MARK: - Status

struct ToolStatusIndicator: View {
    let state: ToolCallState

    var body: some View {
        switch state {
        case .running:
            ProgressView()
                .controlSize(.small)
        case .pending:
            StatusChip(text: String(localized: "pending"), tint: .secondary)
        case .error:
            StatusChip(text: String(localized: "error"), tint: .red)
        case .completed:
            Image(systemName: "checkmark")
                .font(.caption)
                .foregroundStyle(.secondary)
        }
    }
}

private struct StatusChip: View {
    let text: String
    let tint: Color

    var body: some View {
        Text(text)
            .font(.caption2)
            .foregroundStyle(tint)
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(tint.opacity(0.14), in: RoundedRectangle(cornerRadius: 6))
    }
}

// MARK: - Permission (read-only)

/// Read-only permission verdict: highlighted banner while pending (shown only
/// without a `\.chatInteractions` engine — previews/tests; the live chat
/// renders `PendingPermissionFooter` instead), subdued line once decided.
private struct PermissionStateRow: View {
    let permission: ToolPermission

    var body: some View {
        switch permission.status {
        case .pending:
            Label("Awaiting approval", systemImage: "hourglass")
                .font(.footnote.weight(.semibold))
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 10)
                .padding(.vertical, 6)
                .background(.orange.opacity(0.18))
                .foregroundStyle(.orange)
        case .approved:
            PermissionLine(text: String(localized: "✓ Approved") + (permission.mode.map { " · \($0)" } ?? ""))
        case .denied:
            PermissionLine(
                text: String(localized: "✕ Denied") + (permission.reason.map { " · \($0)" } ?? ""),
                isError: true
            )
        case .canceled:
            PermissionLine(text: String(localized: "— Canceled"))
        }
    }
}

private struct PermissionLine: View {
    let text: String
    var isError = false

    var body: some View {
        Text(text)
            .font(.caption2)
            .foregroundStyle(isError ? AnyShapeStyle(.red) : AnyShapeStyle(.secondary))
            .padding(.horizontal, 10)
            .padding(.bottom, 6)
    }
}

File: android/app/src/main/kotlin/app/hapi/companion/feature/chat/blocks/ToolCallBlockView.kt
package app.hapi.companion.feature.chat.blocks

import androidx.compose.animation.animateContentSize
import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import app.hapi.companion.R
import app.hapi.companion.feature.chat.ChatBlockCard
import app.hapi.companion.feature.chat.LocalChatInteractions
import app.hapi.companion.feature.chat.toolCardPresentation
import app.hapi.companion.ui.theme.HapiTheme
import app.hapi.companion.ui.theme.hapi
import app.hapi.protocol.chat.ChatToolCall
import app.hapi.protocol.chat.ToolCallBlock
import app.hapi.protocol.chat.ToolPermission
import app.hapi.protocol.chat.ToolState
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/**
 * One tool invocation (web `ToolCard`): collapsed header row — icon glyph,
 * title, subtitle, status — expanding to the per-tool body ([ToolCallBody]),
 * the read-only permission state, and nested children (sidechain transcript).
 * Cards with a pending permission start expanded and carry the
 * "awaiting approval" banner (actions land in M3b).
 */
@Composable
fun ToolCallBlockView(block: ToolCallBlock, basePath: String?, modifier: Modifier = Modifier) {
    val tool = block.tool
    val resources = LocalContext.current.resources
    val presentation = remember(tool, basePath, resources) { toolCardPresentation(tool, basePath, resources) }
    val pendingPermission = tool.permission?.status == "pending"
    var expanded by rememberSaveable(block.id) { mutableStateOf(pendingPermission) }
    val colors = MaterialTheme.hapi

    Surface(
        shape = RoundedCornerShape(12.dp),
        color = MaterialTheme.colorScheme.surfaceContainerLow,
        modifier = modifier.fillMaxWidth().animateContentSize(),
    ) {
        Column {
            Row(
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable { expanded = !expanded }
                    .padding(horizontal = 10.dp, vertical = 8.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(text = presentation.icon, fontSize = 14.sp)
                Spacer(modifier = Modifier.width(8.dp))
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = presentation.title,
                        style = MaterialTheme.typography.bodyMedium,
                        maxLines = 1,
                        overflow = TextOverflow.Ellipsis,
                    )
                    presentation.subtitle?.let { subtitle ->
                        Text(
                            text = subtitle,
                            style = MaterialTheme.typography.labelSmall,
                            color = colors.hint,
                            maxLines = 1,
                            overflow = TextOverflow.Ellipsis,
                        )
                    }
                }
                Spacer(modifier = Modifier.width(8.dp))
                ToolStatusIndicator(tool.state)
            }

            tool.permission?.let { permission ->
                val interactions = LocalChatInteractions.current
                if (permission.status == "pending" && interactions != null) {
                    Surface(
                        color = MaterialTheme.colorScheme.tertiaryContainer.copy(alpha = 0.35f),
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Column {
                            Text(
                                text = stringResource(R.string.chat_tool_awaiting_approval),
                                style = MaterialTheme.typography.labelLarge,
                                modifier = Modifier.padding(start = 10.dp, top = 6.dp),
                            )
                            PendingPermissionFooter(
                                tool = tool,
                                requestId = permission.id,
                                flavor = interactions.flavor,
                                override = interactions.permissionOverrides[permission.id],
                                onAction = interactions.resolvePermission,
                            )
                        }
                    }
                } else {
                    PermissionStateRow(permission)
                }
            }

            if (expanded) {
                ToolCallBody(
                    tool = tool,
                    basePath = basePath,
                    modifier = Modifier.padding(start = 10.dp, end = 10.dp, bottom = 10.dp),
                )
            }

            if (block.children.isNotEmpty()) {
                ChildrenColumn(block, basePath, expanded)
            }
        }
    }
}

/** Sidechain children, nested behind an indent rail; collapsed to a count row. */
@Composable
private fun ChildrenColumn(block: ToolCallBlock, basePath: String?, parentExpanded: Boolean) {
    var childrenOpen by rememberSaveable("children:" + block.id) { mutableStateOf(parentExpanded) }
    val colors = MaterialTheme.hapi

    val stepsLabel = if (block.children.size == 1) {
        stringResource(R.string.chat_agent_steps_one)
    } else {
        stringResource(R.string.chat_agent_steps_many, block.children.size)
    }
    Text(
        text = (if (childrenOpen) "▾ " else "▸ ") + stepsLabel,
        style = MaterialTheme.typography.labelMedium,
        color = colors.hint,
        modifier = Modifier
            .fillMaxWidth()
            .clickable { childrenOpen = !childrenOpen }
            .padding(horizontal = 10.dp, vertical = 6.dp),
    )
    if (!childrenOpen) return

    Row(
        modifier = Modifier
            .padding(start = 12.dp, end = 8.dp, bottom = 10.dp)
            .height(IntrinsicSize.Min),
    ) {
        Spacer(
            modifier = Modifier
                .width(2.dp)
                .fillMaxHeight()
                .clip(RoundedCornerShape(1.dp))
                .background(colors.divider),
        )
        Column(
            modifier = Modifier
                .weight(1f)
                .padding(start = 10.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            block.children.forEach { child ->
                ChatBlockCard(block = child, basePath = basePath)
            }
        }
    }
}

@Composable
internal fun ToolStatusIndicator(state: String) {
    when (state) {
        ToolState.RUNNING -> CircularProgressIndicator(
            modifier = Modifier.size(14.dp),
            strokeWidth = 2.dp,
        )
        ToolState.PENDING -> StatusChip(
            text = stringResource(R.string.chat_tool_status_pending),
            container = MaterialTheme.colorScheme.surfaceContainerHigh,
            content = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        ToolState.ERROR -> StatusChip(
            text = stringResource(R.string.chat_tool_status_error),
            container = MaterialTheme.colorScheme.errorContainer,
            content = MaterialTheme.colorScheme.onErrorContainer,
        )
        else -> Text(
            text = "✓",
            style = MaterialTheme.typography.labelMedium,
            color = MaterialTheme.hapi.hint,
        )
    }
}

@Composable
private fun StatusChip(text: String, container: androidx.compose.ui.graphics.Color, content: androidx.compose.ui.graphics.Color) {
    Text(
        text = text,
        style = MaterialTheme.typography.labelSmall,
        color = content,
        modifier = Modifier
            .clip(RoundedCornerShape(6.dp))
            .background(container)
            .padding(horizontal = 6.dp, vertical = 2.dp),
    )
}

/**
 * Read-only permission verdict: highlighted banner while pending (renders
 * only without a [LocalChatInteractions] provider — previews/tests; the live
 * chat replaces it with [PendingPermissionFooter]), subdued line once decided.
 */
@Composable
private fun PermissionStateRow(permission: ToolPermission) {
    when (permission.status) {
        "pending" -> Surface(
            color = MaterialTheme.colorScheme.tertiaryContainer,
            contentColor = MaterialTheme.colorScheme.onTertiaryContainer,
            modifier = Modifier.fillMaxWidth(),
        ) {
            Column(modifier = Modifier.padding(horizontal = 10.dp, vertical = 6.dp)) {
                Text(
                    text = stringResource(R.string.chat_tool_awaiting_approval_badge),
                    style = MaterialTheme.typography.labelLarge,
                )
            }
        }
        "approved" -> PermissionLine(
            stringResource(R.string.chat_tool_approved) + (permission.mode?.let { " · $it" } ?: ""),
        )
        "denied" -> PermissionLine(
            stringResource(R.string.chat_tool_denied) + (permission.reason?.let { " · $it" } ?: ""),
            error = true,
        )
        "canceled" -> PermissionLine(stringResource(R.string.chat_tool_canceled))
    }
}

@Composable
private fun PermissionLine(text: String, error: Boolean = false) {
    Text(
        text = text,
        style = MaterialTheme.typography.labelSmall,
        color = if (error) MaterialTheme.colorScheme.error else MaterialTheme.hapi.hint,
        modifier = Modifier.padding(start = 10.dp, end = 10.dp, bottom = 6.dp),
    )
}

// -------------------------------------------------------------- previews --

internal fun previewToolCall(
    id: String,
    name: String,
    state: String = ToolState.COMPLETED,
    input: Map<String, String> = emptyMap(),
    permission: ToolPermission? = null,
): ToolCallBlock = ToolCallBlock(
    id = id,
    localId = null,
    createdAt = 0,
    invokedAt = null,
    tool = ChatToolCall(
        id = id,
        name = name,
        state = state,
        input = JsonObject(input.mapValues { (_, value) -> JsonPrimitive(value) }),
        createdAt = 0,
        description = null,
        permission = permission,
    ),
    children = emptyList(),
    meta = null,
)

@Preview(showBackground = true)
@Composable
private fun ToolCallBlockPreview() {
    HapiTheme {
        Surface {
            Column(
                modifier = Modifier.padding(12.dp),
                verticalArrangement = Arrangement.spacedBy(8.dp),
            ) {
                ToolCallBlockView(
                    previewToolCall("t1", "Bash", ToolState.RUNNING, mapOf("command" to "bun test --watch")),
                    basePath = null,
                )
                ToolCallBlockView(
                    previewToolCall("t2", "Read", input = mapOf("file_path" to "/repo/web/src/chat/reducer.ts")),
                    basePath = "/repo",
                )
                ToolCallBlockView(
                    previewToolCall(
                        "t3",
                        "Bash",
                        ToolState.PENDING,
                        mapOf("command" to "rm -rf build"),
                        permission = ToolPermission(
                            id = "p1",
                            status = "pending",
                            presence = setOf("id", "status"),
                        ),
                    ),
                    basePath = null,
                )
            }
        }
    }
}
